happi/audit.py

```diff
@@ -59,6 +61,60 @@ def check_name_match_id(result: SearchResult) -> None:
     assert id_field == name_field, f'id: {id_field} != name: {name_field}'
 
 
+def check_wait_connection(result: SearchResult) -> None:
+    """
+    Check if all the signals on the device can connect to their PV's
+
+    This can take a long time, consider filtering your query before running
+    """
+    dev = result.get()
+
+    assert (hasattr(dev, 'wait_for_connection') and
+            callable(getattr(dev, 'wait_for_connection'))), \
+           ('device has no wait_for_connection method, and is likely '
+            'not an ophyd device')
+
+    try:
+        dev.wait_for_connection(timeout=5)
+    except TimeoutError as te:
+        # If we encounter a timeout error, gather some more details stats
+        if hasattr(dev, "walk_signals"):
+            sigs = list(sig.item for sig in dev.walk_signals())
+            conn_sigs = sum(sig.connected for sig in sigs)
+
+        raise ValueError(f'{conn_sigs}/{len(sigs)} signals connected. \n'
+                         f'original error: {te}')
+
+
+def check_args_kwargs_match(result: SearchResult) -> None:
+    """
+    Check that arguments that request information from the container
+    have information in those entries.
+
+    i.e.: makes sure there is information in the entry named "extra"
+          if the kwargs = {'extra': '{{extra}}'}
+    """
+    # Happi fills in values when search result is created, must pull
+    # raw document from client level
+    cl = result.client
+    doc = cl.find_document(**{cl._id_key: result[cl._id_key]})
+    # pick out any jinja-like template
+    jinja_pattern = re.compile(r'^\{\{([a-zA-Z][a-zA-Z0-9_]*)\}\}$')
```

Comment:
This expression will work for all common cases, but it's important to remember that this is jinja syntax. There could be things such as:
`{{ a or "foobar" }}` or `{{ a | filter }}`

Jinja, fortunately, has built-in functionality for this: https://github.com/pcdshub/happi/blob/0cb79b5e1f6bb06e1c03e17e22fd316b755ac447/happi/loader.py#L56

Reply:
Ah, I should have thought to look through jinja 🤦